pypit/core/artracewave.py

```diff
@@ -0,0 +1,1235 @@
+
+    # Fit the inverted model with a 2D polynomial
+    msgs.info("Fitting tilts with a low order, 2D {:s}".format(tilt_settings['tilts']['func2D']))
+    wgd = np.where(xtilt != maskval)
+    # Invert
+    coeff2 = arutils.polyfit2d_general(xtilt[wgd], mtilt[wgd]/(msarc.shape[0]-1),
+                                       mtilt[wgd]-ytilt[wgd], fitxy,
+                                              minx=0., maxx=1., miny=0., maxy=1.,
+                                              function=tilt_settings['tilts']['func2D'])
+    polytilts = arutils.polyval2d_general(coeff2, np.linspace(0.0, 1.0, msarc.shape[1]),
+                                          np.linspace(0.0, 1.0, msarc.shape[0]),
+                                          minx=0., maxx=1., miny=0., maxy=1.,
+                                          function=tilt_settings['tilts']['func2D'])
+
+    # TODO -- Add a rejection iteration (or two)
+
+    # Residuals
+    xv2 = arutils.scale_minmax(xtilt[wgd], minx=0., maxx=1)
+    yv2 = arutils.scale_minmax(mtilt[wgd]/(msarc.shape[0]-1), minx=0., maxx=1)
+    yfit = np.polynomial.legendre.legval2d(xv2, yv2, coeff2)
+    res2 = (mtilt[wgd]-ytilt[wgd]) - yfit
+    msgs.info("RMS (pixels): {}".format(np.std(res2)))
+
+    # QA
+    if doqa:
+        plot_tiltres(setup, mtilt[wgd], ytilt[wgd], yfit, slit=slit, show_QA=show_QA)
+
+    # y normalization and subtract
+    ynorm = np.outer(np.linspace(0., 1., msarc.shape[0]), np.ones(msarc.shape[1]))
+    polytilts = ynorm - polytilts/(msarc.shape[0]-1)
+
+    # Return
+    outpar = None
+    return polytilts, outpar
+
+
+def slit_image(slf, det, scitrace, obj, tilts=None):
+    """ Generate slit image for a given object
+    Ignores changing plate scale (for now)
+    The slit is approximated as a straight line in this calculation
+    which should be reasonably accurate.  Better, the object profile
+    generated from this approximation is applied in the same fashion
+    so that the 'error' is compensated for.
+    Parameters
+    ----------
+    slf
+    det
+    scitrace
+    obj
+    Returns
+    -------
+    slit_img : ndarray
+    """
+    # Setup
+    if tilts is None:
+        tilts = slf._tilts[det-1]
+    ximg = np.outer(np.ones(tilts.shape[0]), np.arange(tilts.shape[1]))
+    dypix = 1./tilts.shape[0]
+    #  Trace
+    xtrc = np.round(scitrace['traces'][:, obj]).astype(int)
+    wch = np.where((xtrc >= 0) & (xtrc <= tilts.shape[1]-1))
+    msgs.work("Use 2D spline to evaluate tilts")
+    trc_tilt = np.zeros(tilts.shape[0], dtype=np.float)
+    trc_tilt[wch] = tilts[np.arange(tilts.shape[0])[wch], xtrc[wch]]
+    trc_tilt_img = np.outer(trc_tilt, np.ones(tilts.shape[1]))
+    # Slit image
+    msgs.work("Should worry about changing plate scale")
+    dy = (tilts - trc_tilt_img)/dypix  # Pixels
+    dx = ximg - np.outer(scitrace['traces'][:, obj], np.ones(tilts.shape[1]))
+    slit_img = np.sqrt(dx**2 - dy**2)
```

Comment:
I feel like this routine is assuming that the tilts are modelled as a linear function. Is the idea that they can be assumed linear locally around the object, or am I missing something?

Reply:
I don't think this code is being used.  Have
commented it out.